pilot/pkg/model/push_context.go

```diff
@@ -925,22 +926,41 @@ func (ps *PushContext) DelegateVirtualServices(vses []config.Config) []ConfigHas
 // Callers can check if the sidecarScope is from user generated object or not
 // by checking the sidecarScope.Config field, that contains the user provided config
 func (ps *PushContext) getSidecarScope(proxy *Proxy, workloadLabels labels.Instance) *SidecarScope {
-	// Find the most specific matching sidecar config from the proxy's
-	// config namespace If none found, construct a sidecarConfig on the fly
-	// that allows the sidecar to talk to any namespace (the default
-	// behavior in the absence of sidecars).
-	if sidecars, ok := ps.sidecarIndex.sidecarsByNamespace[proxy.ConfigNamespace]; ok {
-		// TODO: logic to merge multiple sidecar resources
-		// Currently we assume that there will be only one sidecar config for a namespace.
-		if proxy.Type == Router {
+	// TODO: logic to merge multiple sidecar resources
+	// Currently we assume that there will be only one sidecar config for a namespace.
+	sidecars, hasSidecar := ps.sidecarIndex.sidecarsByNamespace[proxy.ConfigNamespace]
+	switch proxy.Type {
+	case Router:
+		if hasSidecar {
+			// Find the most specific matching sidecar config from the proxy's
+			// config namespace If none found, construct a sidecarConfig on the fly
+			// that allows the sidecar to talk to any namespace (the default
+			// behavior in the absence of sidecars).
 			for _, wrapper := range sidecars {
 				// Gateways should just have a default scope with egress: */*
 				if wrapper.Sidecar == nil {
 					return wrapper
 				}
 			}
 		}
-		if proxy.Type == SidecarProxy {
+		// We didn't have a Sidecar in the namespace. This means we should use the default - either an implicit
+		// default selecting everything, or pulling from the root namespace.
+		ps.sidecarIndex.derivedSidecarMutex.RLock()
```

Comment:
lock and defer unlock here

Reply:
I did this because, we can do a read only lock here as that is the most common use case once pod connects to istiod for further pushes. Write lock is only needed on init connects